sympy/polys/domains/quaternion.py

```diff
@@ -0,0 +1,524 @@
+# References :
+# http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/
+# https://en.wikipedia.org/wiki/Quaternion
+
+from __future__ import print_function
+
+from sympy.core.basic import Basic
+from sympy import Rational
+from sympy import re, im
+from sympy import sqrt, sin, cos, acos, asin, exp, ln
+from sympy import trigsimp
+from sympy import diff, integrate
+from sympy import Matrix
+from sympy import symbols
+from sympy.printing.latex import LatexPrinter, latex
+from sympy.printing import StrPrinter
+from sympy import sstr
+
+
+class Quaternion(Basic):
+
+    """Provides basic quaternion operations.
+    Quaternion objects can be instantiated as Quaternion(a, b, c, d)
+    as in (a + b*i + c*j + d*k).
+
+    Example
+    ========
+
+    >>> from sympy.polys.domains.quaternion import Quaternion
+    >>> q = Quaternion(1, 2, 3, 4)
+    >>> print(q)
+    1 + (2)i + (3)j + (4)k
+    """
+
+    def __init__(
+        self,
+        a=0,
+        b=0,
+        c=0,
+        d=0,
+    ):
+
+        self.a = a
```

Comment:
These should be transformed into a SymPy object.

Reply:
Would it be enough to use simpify()?

Reply:
the function `a = sympify(a)`